File: setbot.js
#!/usr/bin/env node

const fs = require('fs').promises;
const { exec } = require('child_process');
const util = require('util');
const execPromise = util.promisify(exec);
const semver = require('semver'); // Added for version checking

// ANSI color codes (used until chalk is installed)
const COLORS = {
  reset: '\x1b[0m',
  cyan: '\x1b[36m',
  yellow: '\x1b[33m',
  red: '\x1b[31m',
};

// Simple logger
const logger = {
  info: (msg) => console.log(`${COLORS.cyan}[${new Date().toISOString()}] [INFO] ${msg}${COLORS.reset}`),
  warn: (msg) => console.log(`${COLORS.yellow}[${new Date().toISOString()}] [WARN] ${msg}${COLORS.reset}`),
  error: (msg) => console.error(`${COLORS.red}[${new Date().toISOString()}] [ERROR] ${msg}${COLORS.reset}`),
};

// List of required npm packages
const PACKAGES = [
  'dotenv',
  'axios',
  'technicalindicators',
  'chalk',
  'semver', // Added for version checking
];

// Default environment variables for .env file
const DEFAULT_ENV = `
BYBIT_API_KEY=your_bybit_api_key_here
BYBIT_API_SECRET=your_bybit_api_secret_here
LOG_LEVEL=INFO
BYBIT_API_BASE_URL=https://api-testnet.bybit.com
`.trim();

// Default configuration for config.json
const DEFAULT_CONFIG = {
  symbol: 'BTC/USDT',
  timeframe: '1h',
  history_candles: 300,
  analysis_interval: 60000,
  max_retries: 5,
  retry_delay: 2000,
  risk_percentage: 0.1,
  max_position_size_usdt: 100,
  stop_loss_multiplier: 2,
  take_profit_multiplier: 4,
  sl_tp_offset_percentage: 0.001,
  cache_ttl: 300000,
  indicators: {
    sma: { fast: 10, slow: 30 },
    stoch_rsi: { rsi_period: 14, stochastic_period: 14, k_period: 3, d_period: 3 },
    atr: { period: 14 },
    macd: { fast_period: 12, slow_period: 26, signal_period: 9 },
  },
  thresholds: { oversold: 20, overbought: 80, min_confidence: 60 },
  volume_confirmation: { enabled: true, lookback: 20 },
  fibonacci_pivots: {
    enabled: true,
    period: '1d',
    proximity_percentage: 0.005,
    order_book_range_percentage: 0.002,
    pivot_weight: 15,
    order_book_weight: 10,
    levels_for_buy_entry: ['S1', 'S2'],
    levels_for_sell_entry: ['R1', 'R2'],
  },
};

async function checkNodeVersion() {
  const requiredVersion = '14.0.0';
  const { stdout } = await execPromise('node -v');
  const currentVersion = stdout.trim().replace('v', '');

  if (semver.lt(currentVersion, requiredVersion)) {
    logger.error(`Node.js version ${currentVersion} detected. Please upgrade to ${requiredVersion} or higher.`);
    process.exit(1);
  } else {
    logger.info(`Node.js version check passed: ${currentVersion}`);
  }
}

async function installPackages() {
  logger.info('Installing required npm packages...');
  for (const pkg of PACKAGES) {
    try {
      await execPromise(`npm list ${pkg} --depth=0`);
      logger.info(`${pkg} is already installed`);
    } catch {
      logger.info(`Installing package: ${pkg}`);
      try {
        const { stderr } = await execPromise(`npm install ${pkg}`);
        if (stderr && !stderr.includes('up to date')) {
          logger.warn(`Warnings during ${pkg} installation: ${stderr}`);
        }
        logger.info(`Successfully installed ${pkg}`);
      } catch (error) {
        logger.error(`Failed to install ${pkg}: ${error.message}`);
        process.exit(1);
      }
    }
  }
}

async function createConfigFile() {
  const configPath = 'config.json';
  try {
    await fs.access(configPath);
    logger.info(`Configuration file '${configPath}' already exists`);
  } catch {
    await fs.writeFile(configPath, JSON.stringify(DEFAULT_CONFIG, null, 2));
    logger.info(`Created default configuration file at '${configPath}'`);
  }
}

async function createEnvFile() {
  const envPath = '.env';
  try {
    await fs.access(envPath);
    logger.info(`'.env' file already exists`);
  } catch {
    await fs.writeFile(envPath, DEFAULT_ENV);
    logger.info(`Created '.env' file with default environment variables`);
  }
}

async function verifySetup() {
  logger.info('Verifying setup...');
  const missingPackages = [];
  for (const pkg of PACKAGES) {
    try {
      await execPromise(`npm list ${pkg} --depth=0`);
    } catch {
      missingPackages.push(pkg);
    }
  }
  if (missingPackages.length > 0) {
    logger.error(`Missing packages: ${missingPackages.join(', ')}`);
    process.exit(1);
  }

  try {
    await fs.access('.env');
  } catch {
    logger.error(`'.env' file not found`);
    process.exit(1);
  }

  try {
    await fs.access('config.json');
  } catch {
    logger.error(`'config.json' file not found`);
    process.exit(1);
  }

  const dotenv = require('dotenv');
  dotenv.config();
  if (!process.env.BYBIT_API_KEY || !process.env.BYBIT_API_SECRET) {
    logger.warn(`API credentials not set in '.env'. Please add them before running the bot.`);
  } else {
    logger.info(`API credentials found in '.env'`);
  }

  logger.info('Setup verification passed');
}

async function runSetup() {
  try {
    logger.info('Starting setup for Trading Bot...');
    await checkNodeVersion();
    await installPackages();
    await createConfigFile();
    await createEnvFile();
    await verifySetup();
    logger.info('Setup completed successfully! You can now run the trading bot.');
    logger.info('Next steps:\n1. Edit \'.env\' with your Bybit API credentials.\n2. Customize \'config.json\' if needed.\n3. Run the bot with \'node trading_bot.js\'');
  } catch (error) {
    logger.error(`Setup failed: ${error.message}`);
    console.error(error.stack);
    process.exit(1);
  }
}

if (require.main === module) {
  runSetup();
}

File: setbot.rb
#!/usr/bin/env ruby

require 'fileutils'
require 'open3'
require 'logger' # Added this line to fix the NameError

# Setup script for the trading bot
class TradingBotSetup
  # List of required Ruby gems
  GEMS = %w[dotenv ccxt technicalindicators colorize async hashie].freeze

  # Default environment variables for .env file
  DEFAULT_ENV = <<~ENV
    BYBIT_API_KEY='your_bybit_api_key_here'
    BYBIT_API_SECRET='your_bybit_api_secret_here'
    LOG_LEVEL='INFO'
  ENV

  def initialize
    @logger = Logger.new(STDOUT)
    @logger.formatter = proc { |severity, datetime, _, msg| "[#{datetime.iso8601}] [#{severity}] #{msg}\n" }
  end

  def run
    @logger.info('Starting setup for Trading Bot...')
    check_ruby_version
    install_gems
    create_config_file
    create_env_file
    verify_setup
    @logger.info('Setup completed successfully! You can now run the trading bot.')
    @logger.info("Next steps:\n1. Edit '.env' with your Bybit API credentials.\n2. Customize 'config.json' if needed.\n3. Run the bot with 'ruby trading_bot.rb'")
  rescue StandardError => e
    @logger.error("Setup failed: #{e.message}")
    @logger.error(e.backtrace.join("\n"))
    exit 1
  end

  private

  def check_ruby_version
    required_version = '3.0.0'
    current_version = RUBY_VERSION
    if Gem::Version.new(current_version) < Gem::Version.new(required_version)
      @logger.error("Ruby version #{current_version} detected. Please upgrade to #{required_version} or higher.")
      exit 1
    else
      @logger.info("Ruby version check passed: #{current_version}")
    end
  end

  def install_gems
    @logger.info('Installing required Ruby gems...')
    GEMS.each do |gem|
      installed = system("gem list -i #{gem} > /dev/null 2>&1")
      unless installed
        @logger.info("Installing gem: #{gem}")
        stdout, stderr, status = Open3.capture3("gem install #{gem}")
        unless status.success?
          @logger.error("Failed to install #{gem}: #{stderr}")
          exit 1
        end
        @logger.info("Successfully installed #{gem}")
      else
        @logger.info("#{gem} is already installed")
      end
    end
  end

  def create_config_file
    config_path = 'config.json'
    unless File.exist?(config_path)
      default_config = {
        symbol: 'BTC/USDT',
        timeframe: '1h',
        history_candles: 300,
        analysis_interval: 60_000,
        max_retries: 5,
        retry_delay: 2000,
        risk_percentage: 0.1,
        max_position_size_usdt: 100,
        stop_loss_multiplier: 2,
        take_profit_multiplier: 4,
        sl_tp_offset_percentage: 0.001,
        cache_ttl: 300_000,
        indicators: {
          sma: { fast: 10, slow: 30 },
          stoch_rsi: { rsi_period: 14, stochastic_period: 14, k_period: 3, d_period: 3 },
          atr: { period: 14 },
          macd: { fast_period: 12, slow_period: 26, signal_period: 9 }
        },
        thresholds: { oversold: 20, overbought: 80, min_confidence: 60 },
        volume_confirmation: { enabled: true, lookback: 20 },
        fibonacci_pivots: {
          enabled: true,
          period: '1d',
          proximity_percentage: 0.005,
          order_book_range_percentage: 0.002,
          pivot_weight: 15,
          order_book_weight: 10,
          levels_for_buy_entry: %w[S1 S2],
          levels_for_sell_entry: %w[R1 R2]
        }
      }
      File.write(config_path, JSON.pretty_generate(default_config))
      @logger.info("Created default configuration file at '#{config_path}'")
    else
      @logger.info("Configuration file '#{config_path}' already exists")
    end
  end

  def create_env_file
    env_path = '.env'
    unless File.exist?(env_path)
      File.write(env_path, DEFAULT_ENV)
      @logger.info("Created '.env' file with default environment variables")
    else
      @logger.info("'.env' file already exists")
    end
  end

  def verify_setup
    @logger.info('Verifying setup...')
    missing_gems = GEMS.reject { |gem| system("gem list -i #{gem} > /dev/null 2>&1") }
    unless missing_gems.empty?
      @logger.error("Missing gems: #{missing_gems.join(', ')}")
      exit 1
    end

    unless File.exist?('.env')
      @logger.error("'.env' file not found")
      exit 1
    end

    unless File.exist?('config.json')
      @logger.error("'config.json' file not found")
      exit 1
    end

    require 'dotenv'
    Dotenv.load
    unless ENV['BYBIT_API_KEY'] && ENV['BYBIT_API_SECRET']
      @logger.warn("API credentials not set in '.env'. Please add them before running the bot.")
    else
      @logger.info("API credentials found in '.env'")
    end

    @logger.info('Setup verification passed')
  end
end

# Run the setup
if $PROGRAM_NAME == __FILE__
  TradingBotSetup.new.run
end
